cron/src/main/scala/com/criteo/cuttle/cron/CronScheduler.scala

```diff
@@ -0,0 +1,155 @@
+                     logger: Logger = logger): Unit = {
+    def runScheduledJob(job: CronJob, retryNum: Int, scheduledAt: ScheduledAt) =
+      for {
+        runInfo <- IO {
+          logger.debug(s"Sending job ${job.underlying.id} to executor")
+          val cronContext = CronContext(scheduledAt.instant)(retryNum)
+          executor.run(job.underlying, cronContext)
+        }
+        _ <- addNextExecutionToState(job, runInfo._1)
+        _ <- logState()
+        result <- IO.fromFuture(IO(runInfo._2))
+      } yield result
+
+    def run(job: CronJob): IO[Completed] = {
+      val maybeNextJob = job match {
+        case newJob @ New(underlying) =>
+          for {
+            maybeScheduledAt <- IO.pure(job.nextEvent)
+            nextJob <- maybeScheduledAt match {
+              case None => IO.pure(None)
+              case Some(scheduledAt) =>
+                logger.debug(
+                  s"Run job ${newJob.underlying.id} at ${scheduledAt.instant} with a delay of ${scheduledAt.delay}")
+                for {
+                  _ <- addNextEventToState(job, scheduledAt.instant)
+                  _ <- logState()
+                  _ <- Timeout.applyF(scheduledAt.delay)
+
+                  nextJob <- runScheduledJob(newJob, 0, scheduledAt)
+                    .map(_ => Some(newJob: CronJob))
+                    .recover {
+                      case _ =>
+                        logger.debug(s"Job ${newJob.underlying.id} has failed it's going to be retried")
+                        Some(Retry(underlying, 1, scheduledAt))
+                    }
+                } yield nextJob
+            }
+            _ <- removeJobFromState(newJob)
+          } yield nextJob
+
+        case retryJob @ Retry(underlying, retryNum, scheduledAt) =>
+          logger.debug(s"Retry job ${retryJob.underlying.id} at ${scheduledAt.instant}")
+          for {
+            nextJob <- runScheduledJob(retryJob, retryNum, scheduledAt)
+              .map(_ => Some(New(underlying): CronJob))
+              .recover {
+                case _ =>
+                  logger.debug(
+                    s"Job ${retryJob.underlying.id} has failed it's going to be retried. Retry number: $retryNum")
+                  Some(Retry(underlying, retryNum + 1, scheduledAt))
+              }
+            _ <- removeJobFromState(retryJob)
+          } yield nextJob
+      }
+
+      maybeNextJob
+        .flatMap {
+          case None =>
+            logger.info(s"Job ${job.underlying.id} has finished. We will not submit executions anymore")
+            IO.pure(Completed)
+          case Some(nextJob) => run(nextJob)
+        }
+        .recover {
+          case e =>
+            val message = s"Fatal error Cron loop for job ${job.underlying.id} will be stopped"
+            logger.error(message)
+            throw new Exception(message)
+        }
+    }
+
+    val programs = workload match {
+      case CronWorkload(jobs) => jobs map New map run
```

Comment:
I'm sorry that's not very idiomatic regarding the rest of the code base.
```suggestion
      case CronWorkload(jobs) => jobs.map(New).map(run)
```

Reply:
Got it